### PR TITLE
fix: [PL-28236]: Editing the stage and pipeline template reseting the step what we have configured

### PR DESCRIPTION
**Pull request by sumanmabd**
### Summary

For pipeline and stage Template, when we add any number of steps, and if we try to edit the template and press continue, the steps that we created will cleaned up. This is fixed.

#### Screenshots

NA

#### PR Checklist

<details>
<summary>Please check if your PR fulfils the following requirements</summary>

- [ ] Tests for the changes have been added. Ideally, include a test that fails without this PR but passes with it.
- [ ] Docs have been [added/updated](https://harness.atlassian.net/jira/software/c/projects/DOC/boards/40).
</details>

<details>
<summary>Use the following comments to re-trigger PR Checks</summary>

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress: `trigger cypress`
- Fix Prettier: `fix prettier`
</details>

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_NA_** 
 


            
 [Details](https://app.embold.io/organization/gh/harness/repositories/e864badd0b95ab81c060b15b85f15701/pull-request/11369)

File: src/modules/72-templates-library/components/TemplateStudio/PipelineTemplateCanvas/PipelineTemplateCanvasWrapper.tsx [hunk 1]
```diff
@@ -39,19 +39,22 @@ const PipelineTemplateCanvasWrapper = (): JSX.Element => {
...
+  const createPipelineFromTemplate = React.useCallback(
+    (): PipelineInfoConfig =>
+      merge({}, template.spec, {
+        name: DefaultNewPipelineName,
+        identifier: DefaultNewPipelineId
+      }),
+    [template.spec]
+  )
 
-  const [pipeline, setPipeline] = React.useState<PipelineInfoConfig>(createPipelineFromTemplate())
```
Comment (depanshu-dhiman) [inline src/modules/72-templates-library/components/TemplateStudio/PipelineTemplateCanvas/PipelineTemplateCanvasWrapper.tsx:48]: Revert this line pls. We are calling `createPipelineFromTemplate`

File: src/modules/72-templates-library/components/TemplateStudio/StageTemplateCanvas/StageTemplateCanvasWrapper.tsx [hunk 2]
```diff
@@ -31,28 +31,32 @@ const StageTemplateCanvasWrapper = () => {
...
+            name: DefaultNewStageName,
+            identifier: DefaultNewStageId
+          })
+        )
+      }),
+    [template.spec]
+  )
 
-  const [pipeline, setPipeline] = React.useState<PipelineInfoConfig>(createPipelineFromTemplate())
+  const [Template, setTemplate] = React.useState<PipelineInfoConfig>(createPipelineFromTemplate())
```
Comment (depanshu-dhiman) [inline src/modules/72-templates-library/components/TemplateStudio/StageTemplateCanvas/StageTemplateCanvasWrapper.tsx:49]: Revert

File: src/modules/72-templates-library/components/TemplateStudio/StageTemplateCanvas/StageTemplateCanvasWrapper.tsx [hunk 3]
```diff
@@ -62,7 +66,7 @@ const StageTemplateCanvasWrapper = () => {
   return (
     <TemplatePipelineProvider
       queryParams={{ accountIdentifier: accountId, orgIdentifier, projectIdentifier }}
-      initialValue={pipeline}
+      initialValue={Template}
```
Comment (depanshu-dhiman) [inline src/modules/72-templates-library/components/TemplateStudio/StageTemplateCanvas/StageTemplateCanvasWrapper.tsx:69]: Revert

File: src/modules/72-templates-library/components/TemplateStudio/StageTemplateCanvas/StageTemplateCanvasWrapper.tsx [hunk 4]
```diff
@@ -22,6 +22,7 @@ import type { ProjectPathProps } from '@common/interfaces/RouteInterfaces'
 import { sanitize } from '@common/utils/JSONUtils'
 import { PipelineContextType } from '@pipeline/components/PipelineStudio/PipelineContext/PipelineContext'
 
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
```
Comment (depanshu-dhiman) [inline src/modules/72-templates-library/components/TemplateStudio/StageTemplateCanvas/StageTemplateCanvasWrapper.tsx:25]: Remove rule

File: src/modules/72-templates-library/components/TemplateStudio/StageTemplateCanvas/StageTemplateCanvasWrapper.tsx [hunk 5]
```diff
@@ -31,28 +32,33 @@ const StageTemplateCanvasWrapper = () => {
...
 
   React.useEffect(() => {
-    if (!isLoading && !isUpdated) {
+    if (!isLoading && isUpdated) {
       setPipeline(createPipelineFromTemplate())
     }
-  }, [isLoading, isUpdated])
+  }, [createPipelineFromTemplate, isLoading, isUpdated])
 
+  // eslint-disable-next-line @typescript-eslint/explicit-function-return-type
```
Comment (depanshu-dhiman) [inline src/modules/72-templates-library/components/TemplateStudio/StageTemplateCanvas/StageTemplateCanvasWrapper.tsx:58]: Remove rule

**Review by depanshu-dhiman — APPROVED** [review]

**Review by yc147 — CHANGES_REQUESTED** [review]
This needs to be testing throughly.

**Pull request closed without merging**